File: study/ts/script/scene/view/tower/TowerAvatarNode.ts
const { ccclass, property } = cc._decorator;

import { Colors, G_EffectGfxMgr, G_Prompt, G_SceneManager, G_UserData } from '../../../init';
import { Lang } from '../../../lang/Lang';
import CommonHeroAvatar from '../../../ui/component/CommonHeroAvatar';
import { handler } from '../../../utils/handler';
import { Path } from '../../../utils/Path';
import UIHelper from '../../../utils/UIHelper';
import PopupTowerChoose from './PopupTowerChoose';

@ccclass
export default class TowerAvatarNode extends cc.Component {

    @property({ type: cc.Node, visible: true })
    _panelTouch: cc.Node = null;
    @property({ type: CommonHeroAvatar, visible: true })
    _panelAvatar: CommonHeroAvatar = null;
    @property({ type: cc.Node, visible: true })
    _nodeInfo: cc.Node = null;
    @property({ type: cc.Label, visible: true })
    _stageName: cc.Label = null;
    @property({ type: cc.Node, visible: true })
    _nodeSword: cc.Node = null;
    @property({ type: cc.Node, visible: true })
    _starPanel3: cc.Node = null;
    @property({ type: cc.Node, visible: true })
    _starPanel2: cc.Node = null;
    @property({ type: cc.Node, visible: true })
    _starPanel1: cc.Node = null;

    private static STAR_DELAY = 0.3
    private static STAR_SCALE = 0.7

    private _layerConfig;
    private _layerData;
    private _nextLayer;
    private _nowLayer;
    private _starEftCount;
    private _starEftFinish;
    private _starEffect: cc.Node[];
    private _starPanel: cc.Node[];

    public init(index) {
        this._layerConfig = null;
        this._layerData = null;
        this._nextLayer = null;
        this._nowLayer = 0;
        this._starEftCount = 0;
        this._starEftFinish = 0;
        this._starEffect = [];

        this._starPanel = [
            this._starPanel1,
            this._starPanel2,
            this._starPanel3
        ];

        this.node.name = "TowerAvatarNode" + (index + 1);
        this._panelTouch.on(cc.Node.EventType.TOUCH_START, handler(this, this._onAvatarClick));
    }

    public refresh(layerData, layerConfig, nextLayer) {
        this._layerConfig = layerConfig;
        this._layerData = layerData;
        this._nextLayer = nextLayer;
        this._nowLayer = G_UserData.getTowerData().getNow_layer();
        this._stageName.string = (this._layerConfig.name);
        this._stageName.node.color = (Colors.getColor(this._layerConfig.color));
        UIHelper.enableOutline(this._stageName, Colors.getColorOutline(this._layerConfig.color), 1);
        this._panelAvatar.updateUI(this._layerConfig.res_id);
        this._panelAvatar.setTouchEnabled(false);
        this._panelAvatar.turnBack();
        var height = this._panelAvatar.getHeight();
        this._nodeInfo.y = (height);
        var layerId = this._layerConfig.id;
        if (layerId <= this._nowLayer) {
            this.showSword(false);
            this.showBubble(false);
        } else if (layerId == this._nextLayer) {
            this.showSword(true);
            this.showBubble(true);
        } else {
            this.showSword(false);
            this.showBubble(false);
        }
        this._refreshStar();
    }

    private _refreshStar() {
        if (this._layerConfig.id == this._nowLayer && G_UserData.getTowerData().isShowStarEft()) {
            this._setStarCount(0);
            this._playStarEft();
            G_UserData.getTowerData().setShowStarEft(false);
            return;
        }
        var star = 0;
        if (this._layerData) {
            star = this._layerData.getNow_star();
        }
        this._setStarCount(star);
    }

    private _setStarCount(count) {
        this._clearStarEffect();
        for (let i in this._starPanel) {
            var v = this._starPanel[i];
            var starNode = v.getChildByName('Star');
            starNode.active = (false);
            if (i < count) {
                starNode.active = (true);
            }
        }
    }

    private _clearStarEffect() {
        for (let i in this._starEffect) {
            var v = this._starEffect[i];
            v.destroy();
        }
        this._starEffect = [];
    }

    public showSword(s) {
        this._nodeSword.removeAllChildren();
        if (s) {
            var effect = G_EffectGfxMgr.createPlayMovingGfx(this._nodeSword, 'moving_shuangjian', null, null, false);
        }
    }

    public showBubble(s) {
        if (s) {
            this._panelAvatar.setBubble(this._layerConfig.talk, null, 2);
        } else {
            this._panelAvatar.setBubbleVisible(false);
        }
    }

    public _onAvatarClick() {
        var layerId = this._layerConfig.id;
        if (layerId == this._nextLayer) {
            if (this._nextLayer == this._nowLayer) {
                G_Prompt.showTip(Lang.get('challenge_tower_already'));
            } else {
                G_SceneManager.openPopup(Path.getPrefab("PopupTowerChoose", "tower"), (popupTowerChoose: PopupTowerChoose) => {
                    popupTowerChoose.init(this._layerConfig);
                    popupTowerChoose.openWithAction();
                });
            }
        } else if (layerId > this._nextLayer) {
            G_Prompt.showTip(Lang.get('challenge_tower_not_reach'));
        } else if (layerId < this._nextLayer) {
            G_Prompt.showTip(Lang.get('challenge_tower_already'));
        }
    }

    private _playStarEft() {
        var star = this._layerData.getNow_star();
        this._starEftCount = star;
        this._starEftFinish = 0;
        for (let i = 0; i < star; i++) {
            var starPanel = this._starPanel[i];
            var delayTime = TowerAvatarNode.STAR_DELAY * (i);
            this._playSingleStarEft(starPanel, delayTime);
        }
    }

    private _playSingleStarEft(node, delayTimme) {
        function eventFunction(event) {
            if (event == 'finish') {
                this._starEftFinish = this._starEftFinish + 1;
                if (this._starEftFinish >= this._starEftCount) {
                }
            }
        }
        function funcStar() {
            var effect = G_EffectGfxMgr.createPlayMovingGfx(node, 'moving_xiaoxingxing', null, eventFunction, false).node;
            var nodeSize = node.getContentSize();
            // var pos = cc.v2(nodeSize.width * 0.5, nodeSize.height * 0.5);
            var pos = cc.v2(0, 0);
            effect.setPosition(pos);
            effect.setScale(TowerAvatarNode.STAR_SCALE);
            this._starEffect.push(effect);
        }
        var action1 = cc.delayTime(delayTimme);
        var action2 = cc.callFunc(funcStar.bind(this));
        var action = cc.sequence(action1, action2);
        node.runAction(action);
    }
}

File: study/lua/src/app/scene/view/tower/TowerAvatarNode.lua
local ViewBase = require("app.ui.ViewBase")
local TowerAvatarNode = class("TowerAvatarNode", ViewBase)

TowerAvatarNode.STAR_DELAY = 0.3
TowerAvatarNode.STAR_SCALE = 0.7

function TowerAvatarNode:ctor(index)
    
    self._panelAvatar = nil     --avatar
    self._panelTouch = nil      --点击面板
    self._nodeInfo = nil        --信息节点
    self._stageName = nil       --节点名字
    self._starPanel1 = nil      --星星1
    self._starPanel2 = nil      --星星2
    self._starPanel3 = nil      --星星3
    self._nodeSword = nil       --剑特效

    self._layerConfig = nil     --表格信息
    self._layerData = nil       --信息
    self._nextLayer = nil       --下一层，需要攻啊的层数
    self._nowLayer = 0          --现在层数

	self._starEftCount = 0 		--播放星星的数量
	self._starEftFinish = 0		--完成播放动画的数量
    self._starEffect = {}

    local resource = {
        file = Path.getCSB("TowerAvatarNode", "tower"),
        size = {1136, 640},
		binding = {
            _panelTouch = {
				events = {{event = "touch", method = "_onAvatarClick"}}
			},
		}
    }
    self:setName("TowerAvatarNode"..index)
    TowerAvatarNode.super.ctor(self, resource)
end

function TowerAvatarNode:onCreate()
    self._starPanel = {self._starPanel1, self._starPanel2, self._starPanel3}
end

--刷新，传入信息，表格，下一层攻打层数
function TowerAvatarNode:refresh(layerData, layerConfig, nextLayer)
    self._layerConfig = layerConfig
    self._layerData = layerData
    self._nextLayer = nextLayer
    self._nowLayer = G_UserData:getTowerData():getNow_layer()
    self._stageName:setString(self._layerConfig.name)
	self._stageName:setColor(Colors.getColor(self._layerConfig.color))
	self._stageName:enableOutline(Colors.getColorOutline(self._layerConfig.color), 1)
	self._panelAvatar:updateUI(self._layerConfig.res_id)
	self._panelAvatar:setTouchEnabled(false)	
	self._panelAvatar:turnBack()
	local height = self._panelAvatar:getHeight()
	self._nodeInfo:setPositionY(height)

    local layerId = self._layerConfig.id
    if layerId <= self._nowLayer then        --已经打过的
        self:showSword(false)
        self:showBubble(false)
    elseif layerId == self._nextLayer then
        self:showSword(true)
        self:showBubble(true)
    else
        self:showSword(false)
        self:showBubble(false)     
    end
    self:_refreshStar()
end

--刷新星星
function TowerAvatarNode:_refreshStar()
    if self._layerConfig.id == self._nowLayer and G_UserData:getTowerData():isShowStarEft() then
		self:_setStarCount(0)
		self:_playStarEft()
        G_UserData:getTowerData():setShowStarEft(false)
		return 
	end
	local star = 0
    if self._layerData then
	    star = self._layerData:getNow_star()
    end    
	self:_setStarCount(star)
end

--设置星星数量
function TowerAvatarNode:_setStarCount(count)
    self:_clearStarEffect()
	for i, v in pairs(self._starPanel) do
		local starNode = v:getSubNodeByName("Star")
        starNode:setVisible(false)
		if i <= count then
			starNode:setVisible(true)
		end
	end
end

--删除星星动画
function TowerAvatarNode:_clearStarEffect()
    for i, v in pairs(self._starEffect) do
        v:removeFromParent()
    end
    self._starEffect = {}   
end

--更新剑动画
function TowerAvatarNode:showSword(s)
    self._nodeSword:removeAllChildren()
    if s then
    	local EffectGfxNode = require("app.effect.EffectGfxNode")
        local function effectFunction(effect)
            if effect == "effect_shuangjian"then
                local subEffect = EffectGfxNode.new("effect_shuangjian")
                subEffect:play()
                return subEffect 
            end
        end
        local effect = G_EffectGfxMgr:createPlayMovingGfx( self._nodeSword, "moving_shuangjian", effectFunction, nil, false )
    end
end

--说话
function TowerAvatarNode:showBubble(s)
    if s then
        self._panelAvatar:setBubble(self._layerConfig.talk, nil, 2)
	else
        self._panelAvatar:setBubbleVisible(false)
    end
end

--点击人物
function TowerAvatarNode:_onAvatarClick()
    local layerId = self._layerConfig.id
    if layerId == self._nextLayer then
    --self._nextLayer == self._nowLayer 只有宝箱没开启时候的特殊状态
        if self._nextLayer == self._nowLayer then
            G_Prompt:showTip(Lang.get("challenge_tower_already"))
        else
            local popupTowerChoose = require("app.scene.view.tower.PopupTowerChoose").new(self._layerConfig)
            popupTowerChoose:openWithAction()
        end
    elseif layerId > self._nextLayer then
        G_Prompt:showTip(Lang.get("challenge_tower_not_reach"))
    elseif layerId < self._nextLayer then
        G_Prompt:showTip(Lang.get("challenge_tower_already"))
    end
end

--播放星星特效
function TowerAvatarNode:_playStarEft()
	local star = self._layerData:getNow_star()
	self._starEftCount = star
	self._starEftFinish = 0
	for i = 1, star do
		local starPanel = self._starPanel[i]
		local delayTime = TowerAvatarNode.STAR_DELAY * (i-1)
		self:_playSingleStarEft(starPanel, delayTime)
	end
end

--播放单个星星动画
function TowerAvatarNode:_playSingleStarEft(node, delayTimme)
	local EffectGfxNode = require("app.effect.EffectGfxNode")
	local function effectFunction(effect)
        if effect == "effect_xiaoxingxing"then
            local subEffect = EffectGfxNode.new(effect)
            subEffect:play()
            return subEffect 
        end
    end
	local function eventFunction(event)
		if event == "finish" then
			self._starEftFinish = self._starEftFinish + 1
			if self._starEftFinish >= self._starEftCount then
				--动画全部完结
				-- G_SignalManager:dispatch(SignalConst.EVENT_STAR_EFFECT_END)
                -- self:_onEventStarEffectEnd()		
            end
		end
	end
	local function funcStar()	
		local effect = G_EffectGfxMgr:createPlayMovingGfx( node, "moving_xiaoxingxing", effectFunction, eventFunction, false )	
		local nodeSize = node:getContentSize()
		local pos = cc.p(nodeSize.width*0.5, nodeSize.height*0.5)
		effect:setPosition(pos)
		effect:setScale(TowerAvatarNode.STAR_SCALE)
        table.insert(self._starEffect, effect)
	end
	local action1 = cc.DelayTime:create(delayTimme)
	local action2 = cc.CallFunc:create(function() funcStar() end)
	local action = cc.Sequence:create(action1, action2)
	node:runAction(action)
end

return TowerAvatarNode
